chainstate/src/detail/tests/normal_tokens.rs

```diff
@@ -0,0 +1,1558 @@
+                    .add_output(TxOutput::new(
+                        OutputValue::Token(TokenData::TokenTransferV1 {
+                            token_id: first_token_id,
+                            amount: (ISSUED_FUNDS * 2).unwrap(),
+                        }),
+                        OutputPurpose::Transfer(Destination::AnyoneCanSpend),
+                    ))
+                    .add_output(TxOutput::new(
+                        OutputValue::Coin((token_min_issuance_fee * 2).unwrap()),
+                        OutputPurpose::Transfer(Destination::AnyoneCanSpend),
+                    ))
+                    .build(),
+            )
+            .build_and_process();
+
+        assert!(matches!(
+            result,
+            Err(BlockError::StateUpdateFailed(
+                ConnectTransactionError::AttemptToPrintMoney(_, _)
+            ))
+        ));
+    })
+}
+
+#[rstest]
+#[trace]
+#[case(Seed::from_entropy())]
+fn test_tokens_reorgs_and_cleanup_data(#[case] seed: Seed) {
+    utils::concurrency::model(move || {
+        const ISSUED_FUNDS: Amount = Amount::from_atoms(1_000_000);
+        let mut rng = make_seedable_rng(seed);
+
+        let mut tf = TestFramework::default();
+        // Issue a new token
+        let genesis_outpoint_id = TestBlockInfo::from_genesis(tf.genesis()).txns[0].0.clone();
+        let block_index = tf
+            .make_block_builder()
+            .add_transaction(
+                TransactionBuilder::new()
+                    .add_input(TxInput::new(
+                        genesis_outpoint_id,
+                        0,
+                        InputWitness::NoSignature(None),
+                    ))
+                    .add_output(TxOutput::new(
+                        OutputValue::Token(TokenData::TokenIssuanceV1 {
+                            token_ticker: b"SOME".to_vec(),
+                            amount_to_issue: ISSUED_FUNDS,
+                            number_of_decimals: 1,
+                            metadata_uri: "https://some_site.some".as_bytes().to_vec(),
+                        }),
+                        OutputPurpose::Transfer(Destination::AnyoneCanSpend),
+                    ))
+                    .build(),
+            )
+            .build_and_process()
+            .unwrap()
+            .unwrap();
+
+        let issuance_block = tf.block(*block_index.block_id());
+        let token_id = token_id(&issuance_block.transactions()[0]).unwrap();
+
+        // Check tokens available in storage
+        let (issuance_block_id, issuance_tx) = tf.chainstate.get_token_detail(token_id).unwrap();
+        assert!(issuance_block.get_id() == issuance_block_id);
+        assert!(issuance_block.transactions()[0].get_id() == issuance_tx.get_id());
+
+        // Cause reorg
+        tf.create_chain(&tf.genesis().get_id().into(), 5, &mut rng).unwrap();
+
+        // Check that tokens not in storage
+        tf.chainstate.get_token_detail(token_id).unwrap_err();
```

Comment:
Where's the check for the issuance transaction?

Reply:
Also, please always test for the particular error. Don't just accept a generic error.

Reply:
Where's the check of the issuance transaction??

Reply:
Added more comments, and update the code. I hope now is better.

Reply:
Added check that reorg happens, and check that tx_index no longer exists in mainchain